File: 2023-06/problem36/LongestPalindrome.cc
#include <iostream>
#include <string>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        string result;
        map<char, vector<int>> m;
        for (int i = 0; i < s.size(); i++) {
            char c = s[i];
            if (m.find(c) == m.end()) {
                vector<int> v;
                v.push_back(i);
                m[c] = v;
            }
            else {
                vector<int> &v = m[c];
                v.push_back(i);
            }
        }
        for (const auto& [c, v] : m) {
            for (int i = 0; i < v.size(); i++) {
                for (int j = i; j < v.size(); j++) {
                    int pos = v[i];
                    int len = v[j] - v[i] + 1;
                    if (len > result.size()) {
                        string ss = s.substr(pos, len);
                        if (isPalindrome(ss)) {
                            result = ss;
                        }
                    }
                }
            }
        }
        return result;
    }

    bool isPalindrome(string s) {
        int len = s.size();
        for (int i = 0; i < len/2; i++) {
            if (s[i] != s[len-i-1]) {
                return false;
            }
        }
        return true;
    }

    string toString(vector<int> v) {
        string s;
        s += "[";
        for (int i = 0; i < v.size(); i++) {
            s += to_string(v[i]);
            if (i < v.size() - 1) {
                s += ",";
            }
        }
        s += "]";
        return s;
    }
};

int main() {
    Solution s;
    string input = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopponmlkjihgfedcba123456789";
    string result = s.longestPalindrome(input);
    cout << "Input string: " << input << endl;
    cout << "Longest palindromic substring: " << result << endl;
    return 0;
}

File: 2023-06/problem42/LongestPalindrome.java
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

public class LongestPalindrome {
    private String input, output = "";
    private Map<Character, List<Integer>> m;

    public LongestPalindrome(String input) {
        this.input = input;
        populate();
        compute();
    }

    private void populate() {
        m = new HashMap<>();
        for (int i = 0; i < input.length(); i++) {
            Character c = input.charAt(i);
            if (!m.containsKey(c)) 
                m.put(c, new ArrayList<Integer>());
            m.get(c).add(i);
        }
    }

    private boolean isPalindrome(String substring) {
        boolean palindrome = true;
        int len = substring.length();
        for (int i = 0; i < len/2; i++) {
            if (substring.charAt(i) != substring.charAt(len-1-i)) {
                palindrome = false;
                break;
            }
        }
        return palindrome;
    }

    private void compute() {
        if (input.isEmpty()) 
            return;
        output = input.substring(0, 1);
        for (Character c : m.keySet()) {
            List<Integer> lst = m.get(c);
            for (int i = 0; i < lst.size(); i++) {
                for (int j = i+1; j < lst.size(); j++) {
                    int begin = lst.get(i);
                    int end = lst.get(j);
                    if (end - begin + 1 > output.length()) {
                        String substring = input.substring(begin, end + 1);
                        if (isPalindrome(substring)) {
                            output = substring;
                        }
                    }
                }
            }
        }
    }

    @Override
    public String toString() {
        return output;
    }

    public static void test(String testCase) {
        System.out.println("Test case: " + testCase);
        LongestPalindrome result = new LongestPalindrome(testCase);
        System.out.println("Longest palindrome: " + result);
    }

    public static void main(String[] args) {
        test("x");
        test("xx");
        test("xyyx");
        test("1xyyx2");
        test("abcdedcba");
        test("xx1abcdedcba1y");
        test("");
        test("  abba");
    }
}
